File: linked-lists/0234-palindrome-linked-list.py
from structures import ListNode

def isPalindrome(head: ListNode) -> bool:
        stack  = []
        fast = head
        slow = head
        while fast and fast.next:
            stack.append(slow.val)
            fast = fast.next.next
            slow = slow.next
            
        # Has odd number of elements, skipping middle one
        if fast:
            slow = slow.next
            
        while slow:
            top = stack.pop()
            
            if top != slow.val:
                return False
            
            slow = slow.next
                
        return True

File: rust/linked-lists/0234-palindrome-linked-list.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn is_palindrome(head: Option<Box<ListNode>>) -> bool {
        let mut stack = vec![];
        let (mut slow, mut fast) = (head.as_ref(), head.as_ref());
        while fast.is_some() && fast.unwrap().next.is_some() {
            stack.push(slow.unwrap().val);
            slow = slow.unwrap().next.as_ref();
            fast = fast.unwrap().next.as_ref().unwrap().next.as_ref();
        }

        if fast.is_some() {
            slow = slow.unwrap().next.as_ref();
        }

        while slow.is_some() {
            let top = stack.pop().unwrap();
            let val = slow.unwrap().val;
            if top != val {return false};

            slow = slow.unwrap().next.as_ref();
        }

        true
    }
}